### PR TITLE
[Do not merge] api: check for 'token_ttl' in helper TokenTTL()

### PR DESCRIPTION
**Pull request by mattlandis**
The Tokenhelper v2 changes moved the ttl for tokens from Data['ttl']
to Data['token_ttl'] this causes the api.Secret.TokenTTL() function
to return a 0 ttl vaule for some tokens.

I encountered this issue when using service account tokens created using gcp rolesets.  I was using the token as part of an oauth2 call and wrapping it in the oauth2 `ReuseTokenSource` which checks the ttl and reuses the token, calling the inner token source when it expired.  I was experiencing 0 token re-uses and calling vault for a fresh token each time it was used.

This seems to be related to the Tokenhelper v2 changes: https://github.com/hashicorp/vault/commit/263b96ef4ec1ece537a52fdce0f5d01669116b29 and
https://github.com/hashicorp/vault-plugin-auth-gcp/pull/68/commits/64e3fed6b97b37c666a6232afcb2d5a8c6dfc288

### CODE REVIEW

File: api/secret.go [hunk 1]
```diff
@@ -260,11 +260,19 @@ func (s *Secret) TokenTTL() (time.Duration, error) {
...
 	}
 
-	if s.Data == nil || s.Data["ttl"] == nil {
+	if s.Data == nil {
 		return 0, nil
 	}
 
-	ttl, err := parseutil.ParseDurationSecond(s.Data["ttl"])
+	var rawTTL interface{}
+	if s.Data["ttl"] != nil {
```
Comment (chrishoffman) [inline api/secret.go]: Can you prefer token_ttl to ttl?

**Comment by jefferai** [comment]
I don't believe this is the right fix. It's putting something gcp plugin specific into the API.

The way the upgrade code should work in the gcp plugin is, if you are writing ttl, you get ttl back in the data (and token_ttl). If you are writing token_ttl, you get that back in the data. If that's not how it's working, it needs to be fixed there.

**Comment by jefferai** [comment]
I've been trying to figure out why this is a problem in the first place, and I think I know what's going on.

TokenTTL is meant to be used on the output of a call to `auth/token/create`. Because that is not an _authenticating_ call but generates a new token for use by the caller of the endpoint, data is returned inside the `Data` block.

It seems the OP has been using this function on the output of the role read endpoint in the GCP plugin. The fact that this ever worked is pure coincidence, but the function is simply not designed to be used with arbitrary endpoints.

**Comment by mattlandis** [comment]
OK.  I want to make sure I understand where we went wrong on this.

We are using `client.Logical().Read()` to read `gcp/token/<tokenpath>`.
https://github.com/hashicorp/vault/blob/master/api/logical.go#L49

This returns an `*api.Secert` which has a function `TokenTTL()` [src](https://github.com/hashicorp/vault/blob/master/api/secret.go#L254).  But the TokenTTL is not intended to be used in this usecase.

Are we using the wrong client?  Is Read() not the correct way to read the secret? Is the return value api.Secret have methods/properties that do not apply in our use case?

**Comment by mattlandis** [comment]
I also don't think it will apply to only gcp.  The [Tokenhelper v2](https://github.com/hashicorp/vault/commit/263b96ef4ec1ece537a52fdce0f5d01669116b29) seems to be intended to help all plugins write common fields (unless the intention is to separate 'ttl' and 'token_ttl' with the intent for them having different meaning and potentially different values).

**Comment by jefferai** [comment]
> We are using client.Logical().Read() to read gcp/token/<tokenpath>.

There is no `token` path in the gcp auth plugin. There _is_ one in the gcp secrets plugin, but you linked to the auth plugin above.

Assuming you meant the gcp secrets plugin, the answer stays the same as what I said before. That function in the API you're referring to is meant specifically for when you are creating a _Vault_ token _directly against the token store_, as in, via one of the `auth/token/` paths. It is not meant for arbitrary plugin response handling. The fact that it ever worked was pure coincidence (and it may well have not been working correctly even if it appeared to do something) just because there happened to be a shared key in the response map.

**Comment by jefferai** [comment]
And to be clear, tokenhelper isn't meant for "arbitrary token from some arbitrary third party". It's a helper library for auth plugins only -- those that allow generating Vault tokens via logging in.

**Comment by jefferai** [comment]
Closing for now.

**Pull request closed without merging**